mgr/dashboard: Hosts Page Service Links Test

adk3798 opened the pull request:
Test links in "Services" column on hosts page properly link to Performance Counters pages

Fixes: https://tracker.ceph.com/issues/41142

Signed-off-by: Adam King <adking@redhat.com>
Signed-off-by: Rafael Quintero <rquinter@redhat.com>

adk3798 commented:
jenkins test dashboard

adk3798 commented:
jenkins test dashboard

Devp00l commented on src/pybind/mgr/dashboard/frontend/e2e/cluster/hosts.po.ts:
```diff
@@ -1,5 +1,46 @@
+      })
+      .then(() => {
+        // if no links were actually tested, share this information
+        if (links_tested === 0) {
```
You should expect that the link count is higher than zero here in order to make sure this test will fail when the test requirements are not met.
```suggestion
        expect(links_tested > 0).toBe(true)
```

Devp00l commented on src/pybind/mgr/dashboard/frontend/e2e/cluster/hosts.po.ts:
```diff
@@ -1,5 +1,46 @@
+          // check is any services links are present
+          services.getText().then((txt) => {
+            // check that text (links) is present in services box
+            if (txt.length !== 0) {
```
Here you can reduce the indent again and improve readability :)
```suggestion
            if (txt.length === 0) {
              return;
            }
```

adk3798 commented:
Update:
 - Changed use of indexOf to includes when checking for a host to be present in table
 - Checks for a host being present and services being present in host were flipped and now stop running running the services link test (and fail) if either of those conditions don't hold. Also increases readability by having less indents.
- Check for at least one link having been tested was moved from an if to an expect. Instead of printing out a message if no links were tested, the test now fails

tspmelo commented on src/pybind/mgr/dashboard/frontend/e2e/cluster/hosts.po.ts:
```diff
@@ -1,5 +1,49 @@
+  // function that checks all services links work for first
+  // host in table (if a host is present)
+  check_services_links() {
+    this.navigateTo();
```
either remove this line or `hosts.e2e-spec.ts:38`.
```suggestion
```

tspmelo commented on src/pybind/mgr/dashboard/frontend/e2e/cluster/hosts.po.ts:
```diff
@@ -1,5 +1,49 @@
+      .getText()
+      .then((hostcount) => {
+        // Only check for services links if there's at least 1 host present
+        if (hostcount.includes('0 total')) {
```
We should always test this. If there are no hosts then we have a problem somewhere.
```suggestion
```

tspmelo commented on src/pybind/mgr/dashboard/frontend/e2e/cluster/hosts.po.ts:
```diff
@@ -1,5 +1,49 @@
+        // check is any services links are present
+        services.getText().then((txt) => {
+          // check that text (links) is present in services box
+          if (txt.length === 0) {
```
Please use expect to validate this.
```suggestion
          expect(txt.length).toBeGreaterThan(0);
```

adk3798 replied:
I like the idea of adding in an expect expect to explicitly test this. However, I'd still like to leave the if block in. The reason being that the next line after the if block "const links = services.all(by.css('a.ng-star-inserted'));" will cause an Async Timeout error if no links are present (it seems to happen when you do a by css search on something that isn't there). Async Timeout errors will not only cause this test to fail but tend to also cause the next few tests to fail even if they should pass. To avoid that in the case of failure I think it would be fine to both have the explicit expect that tests that links are present and then also have the if to avoid the timeout that would occur afterwards if there are no links.

Devp00l replied:
It's also fine to just let this function end at this point as it will return a promise which is resolved in a function that expects if `links_tested` is greater than 0, which in this case will still be 0.

adk3798 commented:
Update:
 - Removed redundant navigateTo from e2e
 - Added explicit test for presence of a host
 - Added expect that explicitly checks that services are present in first host
